Add telemetry markdown file (#6577)
microsoft/FluidFramework: docs/content/start/telemetry.md

---
title: Quick Start
menuPosition: 4
codeCopyButton: true
aliases:

>   - "/docs/get-started/telemetry/"

Author: got it. Thanks for clarifying

> ---

Author: Not really sure if I need to include this, but given that other files in this folder has it, I added it for consistency.

> With thousands of developers using our framework to build their services and applications, telemetry is an essential part of maintaining and troubleshooting the health of the applications, so we are offering just that! We have integrated a way for the developers to set up a telemetry system by allowing them to create custom logging and handling logics to be passed in and called by our Fluid Framework telemetry pipeline.

Reviewer: ```suggestion
With thousands of developers using our framework to build their services and applications, telemetry is an essential part of maintaining and troubleshooting the health of the applications. Developers may set up a telemetry system which allows them to create custom logging and handling logic to be passed in and called by our Fluid Framework telemetry pipeline.
```
nit: tone here sounds more like a blog post about a new feature, whereas I think these docs are intended to be more perennial.  this would be my suggestion to change that tone

Author: good point! Will make the change

> With thousands of developers using our framework to build their services and applications, telemetry is an essential part of maintaining and troubleshooting the health of the applications, so we are offering just that! We have integrated a way for the developers to set up a telemetry system by allowing them to create custom logging and handling logics to be passed in and called by our Fluid Framework telemetry pipeline.

Author: Will go with this version since it sounds more official

## ITelemetryBaseLogger interface

The `ITelemetryBaseLogger` is an interface within the `@fluidframework/common-definitions` package. This interface can be implemented and passed into the `createContainer()` and `getContainer()` methods in `TinyliciousClient` as part of the `TinyliciousContainerConfig` type parameter.

```ts
export interface TinyliciousContainerConfig {
  id: string;
  logger?: ITelemetryBaseLogger;
}
```

```ts
public async createContainer(
  serviceContainerConfig: TinyliciousContainerConfig,
  containerSchema: ContainerSchema,
): Promise<[container: FluidContainer, containerServices: TinyliciousContainerServices]>

> ```

Reviewer 2: I don't know how the docs are presenting the "front door" for devs, but if this section could be made more generic that would be great. Like, if it's not too "bare" of an API to speak of, the real entry point here is in the constructor of `Loader` and `ILoaderProps.logger`. I'd like to see that mentioned here, unless we're really directing people only to the `FooClient`s, in which case I'd present tinylicious as an example, rather than assuming that context.

Author: good point. I elaborated it to include `Loader`. Exposing the users to the APIs will definitely give them a better understanding of how it works under the hood.

> ```

Author: good point. Will rearrange the interface definitions order

### Properties and methods

The interface contains a `supportTags` property and a `send()` method as shown:

```ts
export interface ITelemetryBaseLogger {
  /**
   * An optional boolean which indicates to the user of this interface that tags (i.e. `ITaggedTelemetryPropertyType`
   * objects) are in use. Eventually this will be a required property, but this is a stopgap that allows older hosts
   * to continue to pass through telemetry without trouble (this property will simply show up undefined), while our
   * current logger implementation in `telmetry-utils` handles tags in a separate manner.
   */
  supportsTags?: true;
  send(event: ITelemetryBaseEvent): void;
}
```

- SupportsTags

>   - These tags are generic strings used to classify different events. In a simple logger, all events are untagged and handled the same by your logger's implementation. However, in some scenarios, where some data should be handled separately (e.g. private customer data), then it would be worthwhile to "tag" the event with some identifier.

Reviewer 2: FYI this is probably going away, and tag support will be assumed.  Documentation about tags would move to a comment documenting what ITelemetryBaseEvent is.

- `send()`

>   - The `send()` method is called by the container's telemetry system whenever a telemetry event occurs. This method takes in an ITelemetryBaseEvent type parameter, which is also within the `@fluidframework/common-definitions` package. Given this method is part of an interface, users can implement a custom telemetry logic for the container's telemetry system to execute.

Author: added an `ITelemetryLogger` example

### Adding complexity

> Different levels of logging complexity can be achieved by adding other attributes to the object implementing the `ITelemetryBaseLogger` interface. For example, the `ITelemetryLogger` interface, also within the `@fluidframework/common-definitions` package, broke down telemetry events into different categories, allowing different logging logics for each category. However, it is imperative to ensure that the `send()` method is ultimately called since it is the actual method that is piped to the container's telemetry system and sends the telemetry event.

Reviewer: ```suggestion
Different levels of logging complexity can be achieved by adding other attributes to the object implementing the `ITelemetryBaseLogger` interface. For example, the `ITelemetryLogger` interface, also within the `@fluidframework/common-definitions` package, breaks down telemetry events into different categories, allowing different logging logic for each category. However, it is imperative to ensure that the `send()` method is ultimately called since it is the actual method that is piped to the container's telemetry system and sends the telemetry event.
```

Reviewer 2: The point about `ITelemetryBaseLogger` v. other enrichment is that _between layers of the Fluid Framework_ only `ITelemetryBaseLogger` should be used.

> Different levels of logging complexity can be achieved by adding other attributes to the object implementing the `ITelemetryBaseLogger` interface. For example, the `ITelemetryLogger` interface, also within the `@fluidframework/common-definitions` package, broke down telemetry events into different categories, allowing different logging logics for each category. However, it is imperative to ensure that the `send()` method is ultimately called since it is the actual method that is piped to the container's telemetry system and sends the telemetry event.

Author: I think this version is a lot more succinct 😀

> Different levels of logging complexity can be achieved by adding other attributes to the object implementing the `ITelemetryBaseLogger` interface. For example, the `ITelemetryLogger` interface, also within the `@fluidframework/common-definitions` package, broke down telemetry events into different categories, allowing different logging logics for each category. However, it is imperative to ensure that the `send()` method is ultimately called since it is the actual method that is piped to the container's telemetry system and sends the telemetry event.

Author: I added some more details, encompassing @markfields comment above.

## ITelemetryBaseEvent interface

> A telemetry event is any errors, performance, and informational (non-error) related events. An event is captured and labeled within the `ITelemetryBaseEvent` parameter mentioned previously. The `ITelemetryBaseEvent` is also an interface and implements the `ITelemetryProperties` type. This interface is the base interface for logging telemetry statements, allowing the user to have any number of properties and will serialize it as a JSON payload. With that said, the interface has 2 properties defined already, `eventName` and `category`. These 2 properties are used by the telemetry system to label and define the telemetry event that has occurred.

Reviewer: ```suggestion
A telemetry event is any error, performance, or informational (non-error) event. An event is captured and labeled within the `ITelemetryBaseEvent` parameter mentioned previously. The `ITelemetryBaseEvent` is also an interface and implements the `ITelemetryProperties` type. This interface is the base interface for logging telemetry statements, allowing the user to have any number of properties and will serialize it as a JSON payload. With that said, the interface has 2 properties defined already, `eventName` and `category`. These 2 properties are used by the telemetry system to label and define the telemetry event that has occurred.
```

Reviewer 2: From my comment above - this is where we need to explain tagging. `ITelemetryProperties` is a type with a string index signature, where the values are either tagged or untagged primitives (string, boolean, number), so when implementing `ITelemetryBaseLogger` the dev needs to how to interpret tagged values.

Author: I added some elaboration on the tagging but I feel like my explanation could use some improvement.

```ts
/**
 * Base interface for logging telemetry statements.
 * Can contain any number of properties that get serialized as json payload.
 * @param category - category of the event, like "error", "performance", "generic", etc.
 * @param eventName - name of the event.
 */
export interface ITelemetryBaseEvent extends ITelemetryProperties {
  category: string;
  eventName: string;
}
```

### Category

Currently, there are 3 categories used by the telemetry system:

- errors - used to classify known errors, e.g. duplicate data store IDs.

> - performance - used to track metrics, e.g. used by the summarizer to track how long it takes to do or load a summary (if it's takes too long then even though its a performance event it'll be considered an error).

Reviewer: ```suggestion
- performance - used to track metrics, e.g. used by the summarizer to track how long it takes to do or load a summary. A performance event will be considered an error if it takes too long.
```

> - generic - used as a catchall for events that are mostly harmless.

Reviewer 2: ```suggestion
- generic - used as a catchall for events that are informational and don't represent an activity with a duration like a performance event.
```

### EventName

> This property is currently used by the telemetry system to indicate the event in a more descriptive manner.

Author: theoretically it should be unique, but we aren't checking if it actually is.

### Adding complexity

> Similar to the `ITelemetryBaseLogger` interface mentioned above, different levels of event complexity can also be achieved by adding other attributes to the object implementing the `ITelemetryBaseEvent` interface. Below are some examples:

Author: What I'm trying to say is that the current `ITelemetryBaseEvent` interface only has `category` and `eventName` properties, so if developers want more properties, they can implement the interface and add to it. Similar to something like this 
`
const newEvent: ITelemetryBaseEvent = {
    ...event,
    category: event.category ?? (error === undefined ?  "generic" : "error"),
  };
`

```ts
if (chunk.version !== undefined) {
  logger.send({
      eventName: "MergeTreeChunk:serializeAsMinSupportedVersion",
      category: "generic",
      fromChunkVersion: chunk.version,
      toChunkVersion: undefined,
  });
}
```

The code snippet here implements a telemetry event without adding much complexity. The telemetry event object here adds only a few other attributes to the `ITelemetryBaseEvent`.

```ts
/**
 * Send a telemetry event with the logger
 *
 * @param event - the event to send
 * @param error - optional error object to log
 */
public sendTelemetryEvent(event: ITelemetryGenericEvent, error?: any) {
  const newEvent: ITelemetryBaseEvent = {
    ...event,
    category: event.category ?? (error === undefined ?  "generic" : "error"),
  };
  if (error !== undefined) {
    TelemetryLogger.prepareErrorObject(newEvent, error, false);
  }
  this.send(newEvent);
}
```

> However, the code snippet here shows a telemetry event object that adds much more complex logics to determine the event that has occurred, then passed the resulting event to the `send()` method for the container's telemetry system to output.

Reviewer: ```suggestion
However, the code snippet here shows a telemetry event object that adds much more complex logic to determine the event that has occurred, then passes the resulting event to the `send()` method for the container's telemetry system to output.
```

Here is the above telemetry event object in action:

```ts

> // Adding this event temporarily so that we can get help debugging if something goes wrong.

Reviewer: should remove this comment since it's not relevant to the example

Author: removed! 😃

this.logger.sendTelemetryEvent({
  eventName: "connectedStateRejected",
  source,
  pendingClientId: this.pendingClientId,
  clientId: this.clientId,
  hasTimer: this.prevClientLeftTimer.hasTimer,
  inQuorum: protocolHandler !== undefined && this.pendingClientId !== undefined
      && protocolHandler.quorum.getMember(this.pendingClientId) !== undefined,
});
```

## Code example

> With the interface already hooked up to the container's telemetry system, it is easy for users to write a custom telemetry object by implementing the `ITelemetryBaseLogger` interface and defining the `send()` method. Below is an example custom telemetry logger, `ConsoleLogger`, that implements the `ITelemetryBaseLogger` interface. As the name suggest, the `ConsoleLogger` defined the `send()` method to stringify the entire event object and print it out.

Reviewer 3: ```suggestion
With the interface already hooked up to the container's telemetry system, it is easy for users to write a custom telemetry object by implementing the `ITelemetryBaseLogger` interface and defining the `send()` method. Below is an example custom telemetry logger, `ConsoleLogger`, that implements the `ITelemetryBaseLogger` interface. As the name suggests, the `ConsoleLogger` defined the `send()` method to stringify the entire event object and print it to the browser console.
```

    constructor() {}
    send(event: ITelemetryBaseEvent) {
        console.log("Custom telemetry object array: ".concat(JSON.stringify(event)));
    }
}
```

This custom logger is then created and passed into the `getContainer()` and `createContainer()` in a `TinyliciousClient` object. The custom logger is now hooked up to the container's telemetry system.

```ts
async function start(): Promise<void> {
    // Create a custom ITelemetryBaseLogger object to pass into the Tinylicious container
    // and hook to the Telemetry system
    const consoleLogger: ConsoleLogger = new ConsoleLogger();

    // Get or create the document depending if we are running through the create new flow
    const client = useFrs ? FrsClient :  new TinyliciousClient();
    const [fluidContainer, containerServices] = createNew
        ? await client.createContainer({ id: containerId, logger: consoleLogger }, containerSchema)
        : await client.getContainer({ id: containerId, logger: consoleLogger }, containerSchema);
```

Now, whenever a telemetry event is encountered, the custom `send()` method gets called and will print out the entire event object.

> ![ConsoleLogger_telemetry_in_action](/images/ConsoleLogger_telemetry_in_action.png "ConsoleLogger_telemetry_in_action")

Author: the image isn't showing up on Github, but so are other images in other markdowns.

---
title: Collecting telemetry
---

Telemetry is an essential part of maintaining the health of modern applications. Fluid Framework provides a way to plug in your own logic to handle telemetry events sent by Fluid. This enables you to integrate the Fluid telemetry along with your other telemetry, and route the event data in whatever way you need.

## ITelemetryBaseLogger interface

> The `ITelemetryBaseLogger` is an interface within the `@fluidframework/common-definitions` package. This interface can be implemented and passed into the client's `createContainer()` and `getContainer()` methods via the config parameter, - Both `FrsContainerConfig` and `TinyliciousContainerConfig` have a member `logger?: ITelemetryBaseLogger` . Both `createContainer()` and `getContainer()` methods will create an instance of a `Loader` class object, where the logger defined in the `[client]ContainerConfig` is passed in as an optional parameter, `ILoaderProps.logger`, in the `Loader` constructor argument.

Reviewer 3: ```suggestion
The `ITelemetryBaseLogger` is an interface within the `@fluidframework/common-definitions` package. This interface can be implemented and passed into the client's `createContainer()` and `getContainer()` methods via the config parameter.

All Fluid service clients (for example, `FrsContainerConfig` and `TinyliciousContainerConfig`) allow passing a `logger?: ITelemetryBaseLogger` into the service client configuration. Both `createContainer()` and `getContainer()` methods will create an instance of a `Loader` class object, where the logger defined in the service client configuration is passed in as an optional parameter, `ILoaderProps.logger`, in the `Loader` constructor argument.
```

[`TinyliciousContainerConfig`](https://github.com/microsoft/FluidFramework/blob/7c030344dd54de0cd88fdceabb63f2d0b8e30860/experimental/framework/tinylicious-client/src/interfaces.ts#L9) interface definition takes an optional parameter `logger`. (The definition is similar to `FrsContainerConfig` interface)

> [`createContainer()`](https://github.com/microsoft/FluidFramework/blob/7c030344dd54de0cd88fdceabb63f2d0b8e30860/experimental/framework/tinylicious-client/src/TinyliciousClient.ts#L50) interface definition takes a `TinyliciousContainerConfig` type as its `serviceContainerConfig` argument. (The `FrsClient` will take `FrsContainerConfig` for its respective methods)

Reviewer 3: You probably want the links to point to main, not commit 7c030344dd54de0cd88fdceabb63f2d0b8e30860.

```ts
/**
 * Send a telemetry event with the logger
 *
 * @param event - the event to send
 * @param error - optional error object to log
 */
public sendTelemetryEvent(event: ITelemetryGenericEvent, error?: any) {
  const newEvent: ITelemetryBaseEvent = {
    ...event,
    category: event.category ?? (error === undefined ?  "generic" : "error"),
  };
  if (error !== undefined) {
    TelemetryLogger.prepareErrorObject(newEvent, error, false);
  }
  this.send(newEvent);
}
```
Like demonstrated here, it is imperative to ensure `send()` is ultimately called at the end of custom properties for the information to be piped to the container's telemetry system and sends the telemetry event.

## ITelemetryBaseEvent interface

All Fluid telemetry events are sent as `ITelemetryBaseEvent`s via the `send()` method in `ITelemetryBaseLogger`. This interface can be augmented, allowing you to add additional properties that will be serialized as JSON. The default required properties, `eventName` and `category`, are set by the telemetry system.

```ts
export interface ITelemetryBaseEvent extends ITelemetryProperties {
  category: string;
  eventName: string;
}
```

> The `ITelemetryBaseEvent` contains `category` and `eventName` properties for labeling and defining a telemetry event.-

Reviewer 3: ```suggestion
The `ITelemetryBaseEvent` contains `category` and `eventName` properties for labeling and defining a telemetry event.
```

Author: thanks for spotting my fat finger 😂

---
title: Collecting telemetry
---

Telemetry is an essential part of maintaining the health of modern applications. Fluid Framework provides a way to plug in your own logic to handle telemetry events sent by Fluid. This enables you to integrate the Fluid telemetry along with your other telemetry, and route the event data in whatever way you need.

## ITelemetryBaseLogger interface

> The `ITelemetryBaseLogger` is an interface within the `@fluidframework/common-definitions` package. This interface can be implemented and passed into the client's `createContainer()` and `getContainer()` methods in both `FrsContainerConfig` and `TinyliciousContainerConfig` type parameters. Both `createContainer()` and `getContainer()` methods will create an instance of a `Loader` class object, where the logger defined in the `[client]ContainerConfig` is passed in as an optional parameter, `ILoaderProps.logger`, in the `Loader` constructor argument.

Reviewer 2: What do the words "type parametes" mean in the 2nd sentence?

Reviewer 2: Ok I get it now.  I might just say:

```suggestion
The `ITelemetryBaseLogger` is an interface within the `@fluidframework/common-definitions` package. This interface can be implemented and passed into the client's `createContainer()` and `getContainer()` methods via the config parameter, - Both `FrsContainerConfig` and `TinyliciousContainerConfig` have a member `logger?: ITelemetryBaseLogger` . Both `createContainer()` and `getContainer()` methods will create an instance of a `Loader` class object, where the logger defined in the `[client]ContainerConfig` is passed in as an optional parameter, `ILoaderProps.logger`, in the `Loader` constructor argument.
```

Author: yeah I think this sums it up better. 😀

```ts
export interface TinyliciousContainerConfig {
  id: string;
  logger?: ITelemetryBaseLogger;
}
```
`TinyliciousContainerConfig` interface definition takes an optional parameter `logger`. (The definition is similar to `FrsContainerConfig` interface)

```ts
public async createContainer(
  serviceContainerConfig: TinyliciousContainerConfig,
  containerSchema: ContainerSchema,
): Promise<[container: FluidContainer, containerServices: TinyliciousContainerServices]>
```
`createContainer()` interface definition takes a `TinyliciousContainerConfig` type as its `serviceContainerConfig` argument. (The `FrsClient` will take `FrsContainerConfig` for its respective methods)

```ts
static async getContainer(
  serviceConfig: TinyliciousContainerConfig,
  objectConfig: ContainerSchema,
): Promise<[container: FluidContainer, containerServices: TinyliciousContainerServices]>
```

> `getContainer()` interface definition takes a `TinyliciousContainerConfig` type as its `serviceContainerConfig` argument. (The `FrsClient` will take `FrsContainerConfig` for its respective methods)

Reviewer 2: Might not be necessary to include these code snippets, it's pretty self-explanatory, and will likely get out of date.

The `TinyliciousContainerConfig` snippet likely will too, so maybe pull that out too, but that's less confusing if it is out of date (you could even put a `...` line before the closing bracket)

Author: updated with permalinks

```ts
const loader = new Loader({
  urlResolver: this.urlResolver,
  documentServiceFactory: this.documentServiceFactory,
  codeLoader,
  logger: tinyliciousContainerConfig.logger,
});
```

> The `Loader` object is called by both `createContainer()` and `getContainer()`, and requires a `ILoaderProps` interface as its constructor argument. `ILoaderProps` interface has an optional logger parameter that will take the `ITelemetryBaseLogger` defined by the user.

Reviewer 2: ```suggestion
The `Loader` constructor is called by both `createContainer()` and `getContainer()`, and requires a `ILoaderProps` interface as its constructor argument. `ILoaderProps` interface has an optional logger parameter that will take the `ITelemetryBaseLogger` defined by the user.
```

```ts
constructor(loaderProps: ILoaderProps) {
  const scope = { ...loaderProps.scope };
  if (loaderProps.options?.provideScopeLoader !== false) {
      scope.ILoader = this;
  }

  this.services = {
      urlResolver: createCachedResolver(MultiUrlResolver.create(loaderProps.urlResolver)),
      documentServiceFactory: MultiDocumentServiceFactory.create(loaderProps.documentServiceFactory),
      codeLoader: loaderProps.codeLoader,
      options: loaderProps.options ?? {},
      scope,
      subLogger: DebugLogger.mixinDebugLogger("fluid:telemetry", loaderProps.logger, { all:{loaderId: uuid()} }),
      proxyLoaderFactories: loaderProps.proxyLoaderFactories ?? new Map<string, IProxyLoaderFactory>(),
      detachedBlobStorage: loaderProps.detachedBlobStorage,
  };
  this.logger = ChildLogger.create(this.services.subLogger, "Loader");
}

> ```

Reviewer 2: Like above, I would avoid pasting huge chunks of code in here that is totally subject to change.  Instead, if you want the reader to be able to dive in and see _how_ this all works, try permalinks to code, like this:  https://github.com/microsoft/FluidFramework/blob/main/packages/loader/container-loader/src/loader.ts#L309

(you can also do multi-line by manually doing something like #L1-L10 in the URL)

Reviewer 2: Oooh I didn't know it would render like that in the comment.  The URL is just: `https://github.com/microsoft/FluidFramework/blob/24f5cd9acd3a7ef83604edd9863a47ea50c30569/packages/loader/container-loader/src/loader.ts#L309`

Reviewer 2: Just click on a line number when viewing a file on github.com, then under the `...` do "Copy permalink"

Author: Did not know Github has permalink built in, been doing it manually 😅 Thanks!!

Reviewer 3: Rather than linking to source code, I would link to the API docs, like the ones here: https://fluidframework.com/docs/apis/common-definitions/itelemetrybaselogger/

`ILoaderProps.logger` is used by `Loader` to pipe to container's telemetry system.

### Properties and methods

The interface contains a `supportTags` property and a `send()` method as shown:

```ts
export interface ITelemetryBaseLogger {

>   supportsTags?: true;

Reviewer 2: It's now officially deprecated in main (still needs to be backported to previous release, but happening soon):  https://github.com/microsoft/FluidFramework/blob/24f5cd9acd3a7ef83604edd9863a47ea50c30569/common/lib/common-definitions/src/logger.ts#L43-46

Author: sure, I'll remove it.

  send(event: ITelemetryBaseEvent): void;
}
```

- supportsTags

>   - Tags are strings used to classify different events. In a simple logger, all events are untagged and handled the same by your logger's implementation. However, in some scenarios, where some data should be handled separately (for example, private customer data), those events could be tagged with a unique string so they could be treated differently downstream.

Reviewer 2: Just cut/paste this paragraph down to the section about `ITelemetryBaseEvent` to give context about the tagged property type.

- `send()`
  - The `send()` method is called by the container's telemetry system whenever a telemetry event occurs. This method takes in an ITelemetryBaseEvent type parameter, which is also within the `@fluidframework/common-definitions` package. Given this method is part of an interface, users can implement a custom telemetry logic for the container's telemetry system to execute.

> ### Custom properties

Reviewer 2: I wouldn't call this section "Custom properties" - that makes me think of putting custom properties on the event being logged, which is a basic feature of how `ITelemetryBaseEvent` works.

This seems to be just pointing out that when you implement `ITelemetryBaseLogger' you may expose a richer API to consumers, but ultimately you'll probably want to call the core `send` function.  In fact, you don't have too, someone could totally implement `ITelemetryLogger` and duplicate the core "sending" logic.  Or do different things.

I guess in the end, `ITelemetryLogger` is an internal implementation of a more interesting logging API and I'm not entirely sure it belongs here.

Author: I agree with having `ITelemetryLogger` as example may defeat our argument of ensuring `send()` is called, but currently I don't see any other code snippets that can be use as example for richer API exposure, which is the primary intention here. 🤔
